### PR TITLE
Updates

### PR DESCRIPTION
**Pull request by prolic**
- Repository has saveAggregateRoot method
- SnapshotStore is an interface
- InMemorySnapshotStore
- update tests
- add docs

### CODE REVIEW

**Comment by codeliner** [comment]
:+1:

File: docs/repositories.md [hunk 1]
```diff
@@ -0,0 +1,257 @@
...
+    );
+
+    $this->assertNotSame($fetchedUser, $fetchedUser2);
+
+    $this->assertEquals('Max Mustermann', $fetchedUser2->name());
+}
+```
+
+Here we first add the user, then load it with the help of the repository and finally we change the user entity.
+The change causes a `UserNameChanged` event. You may noticed that after changing the user name, the user is not passed to
```
Comment (lucascourot) [inline docs/repositories.md]: `notice`

File: docs/repositories.md [hunk 2]
```diff
@@ -0,0 +1,257 @@
...
+a repository method. Only `$this->eventStore->commit();` is called. But as you can see in the last test assertion the username
+is changed and the appropriate domain event was added to the `event_stream`. This happens becasue the repository manages an identity map
+internally. Each aggregate root loaded via `AggregateRepository::getAggregateRoot` is added to the identity map and
+new events recorded by such an agggregate root are added automatically to the event stream on `EventStore::commit`.
+
+**But** the identity map is cleared after each transaction commit. You may noticed the `assertNotSame` checks in the test.
```
Comment (lucascourot) [inline docs/repositories.md]: `notice`

File: docs/repositories.md [hunk 3]
```diff
@@ -0,0 +1,257 @@
...
+The repository keeps an aggregate only in memory as long as the transaction is active. Otherwise multiple long-running
+processes dealing with the same aggregate would run into concurrency issues very fast.
+
+The test case has some more tests including snapshot usage and working with different stream names / strategies.
+Just browse through the test methods for details.
+
+## Loading of thousands aggregates
+If you need to load thousands of aggregates for reading only, your memory can be exhausted, because the 
+`AggregateRepository` uses an identity map. So every loaded aggregate is stored there, unless a commit is executed. If
+you have a read only process, you should consider to clear the identity map at sometime. This can be done by calling
```
Comment (lucascourot) [inline docs/repositories.md]: `some time`

File: docs/repositories.md [hunk 4]
```diff
@@ -0,0 +1,257 @@
...
+$thousandsOfAggregateIds = [];
+$number = count($thousandsOfAggregateIds);
+
+foreach ($thousandsOfAggregateIds as $aggregateId) {
+    $aggregate = $this->repository->getAggregateRoot($aggregateId);
+
+    // do something with the aggregate data e.g. build read model
+
+    // clear on every 500th aggregate
+    if (0 === $number % 500) {
```
Comment (lucascourot) [inline docs/repositories.md]: I don't see the `$number` being incremented.

File: tests/Mock/FaultyAggregateRoot2.php [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+                return 'id';
+            }
+
+            public function popRecordedEvents(): void
+            {
+            }
+
+            /**
+             * @param $event
+             */
```
Comment (lucascourot) [inline tests/Mock/FaultyAggregateRoot2.php]: useless

File: tests/Mock/FaultyAggregateRoot2.php [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+        return '0';
+    }
+
+    public function popRecordedEvents(): void
+    {
+    }
+
+    /**
+     * @param $event
+     */
```
Comment (lucascourot) [inline tests/Mock/FaultyAggregateRoot2.php]: useless

File: tests/Mock/FaultyAggregateRoot.php [hunk 7]
```diff
@@ -0,0 +1,89 @@
...
+             */
+            public function replay($event): void
+            {
+            }
+        };
+    }
+
+    /**
+     * @return string
+     */
```
Comment (lucascourot) [inline tests/Mock/FaultyAggregateRoot.php]: useless

File: tests/Mock/DefaultAggregateRootContract.php [hunk 8]
```diff
@@ -0,0 +1,34 @@
...
+    public function getId(): string;
+
+    /**
+     * @return Message[]
+     */
+    public function popRecordedEvents(): array;
+
+    /**
+     * @param $event
+     */
```
Comment (lucascourot) [inline tests/Mock/DefaultAggregateRootContract.php]: useless

File: tests/Mock/DefaultAggregateRoot.php [hunk 9]
```diff
@@ -0,0 +1,65 @@
...
+     * @return Message[]
+     */
+    public function popRecordedEvents(): array
+    {
+        // not required for this mock
+    }
+
+    /**
+     * @param $event
+     */
```
Comment (lucascourot) [inline tests/Mock/DefaultAggregateRoot.php]: useless

File: docs/repositories.md [hunk 10]
```diff
@@ -0,0 +1,257 @@
...
+ * @param null|callable $messageToEventCallback
+ */
+public function __construct(
+    $identifierMethodName = null,
+    $versionMethodName = null,
+    $popRecordedEventsMethodName = null,
+    $replayEventsMethodsName = null,
+    $staticReconstituteFromHistoryMethodName = null,
+    $eventToMessageCallback = null,
+    $messageToEventCallback = null)
```
Comment (lucascourot) [inline docs/repositories.md]: shouldn't the code examples be in php 7.1?

**Comment by prolic** [comment]
updated develop branch because of feedback provided by @lucascourot 
Thank you very much!

**Pull request merged**